kv.go

```diff
@@ -158,25 +159,30 @@ func decodeHook(fromType reflect.Type, toType reflect.Type, data interface{}) (i
 
 			return dataOutput, nil
 		} else if fromType.Kind() == reflect.String {
-			return gzipReader(data)
+			return readCompressedData(data.(string), gzipReader, base64Reader)
 		}
 	}
 	return data, nil
 }
 
-func gzipReader(data interface{}) (interface{}, error) {
-	buffer := bytes.NewBuffer([]byte(data.(string)))
-	reader, err := gzip.NewReader(buffer)
-	if err != nil {
-		// Try to decode data as Base64 encoded string to assume backward compatibility
-		return base64Reader(data.(string))
+func readCompressedData(data string, fs ...func(io.Reader) (io.Reader, error)) ([]byte, error) {
+	var err error
+	for _, f := range fs {
+		var reader io.Reader
+		reader, err = f(bytes.NewBufferString(data))
+		if err == nil {
+			return ioutil.ReadAll(reader)
+		}
 	}
-	defer reader.Close()
-	return ioutil.ReadAll(reader)
+	return nil, err
+}
+
+func base64Reader(r io.Reader) (io.Reader, error) {
+	return base64.NewDecoder(base64.StdEncoding, r), nil
 }
 
-func base64Reader(data string) ([]byte, error) {
-	return base64.StdEncoding.DecodeString(data)
+func gzipReader(r io.Reader) (io.Reader, error) {
```

Comment:
You don't need that, just pass `gzip.NewReader` when calling `readCompressedData` 😉